src/autowiring/DispatchQueue.cpp

```diff
@@ -67,6 +67,31 @@ void DispatchQueue::DispatchEventUnsafe(std::unique_lock<std::mutex>& lk) {
   (*thunk)();
 }
 
+void DispatchQueue::TryDispatchEventUnsafe(std::unique_lock<std::mutex>& lk) {
+  // Pull the ready thunk off of the front of the queue and pop it while we hold the lock.
+  // Then, we will excecute the call while the lock has been released so we do not create
+  // deadlocks.
+  DispatchThunkBase* pThunk = m_pHead;
+  m_pHead = pThunk->m_pFlink;
+  lk.unlock();
+
+  try { (*pThunk)(); }
+  catch (...) {
+    // Failed to execute thunk, put it back
+    lk.lock();
+    pThunk->m_pFlink = m_pHead;
```

Comment:
From reading the comments in the header, I thought we are putting this thunk back to the tail of the queue?